### PR TITLE
Ncs configuration unification

### PR DESCRIPTION
**Pull request by tejlmand**
This PR covers a couple of patterns that has been identified to be lacking in our current system, such as:

- A board specific configuration folder with only settings for that board.
  Reason, having several config files for boards, child images, partition manager directly in app root leads to messy folder / conf structure for more complex applications, see code snippet 1.
- Static partition manager files for specific board.
  Reason: sometimes a static pm conf is needed only for a given board, but there is no way today to automatically source such file, leading to boilerplate code in `<sample>/CMakeLists.txt`, see code snippet 2.
- Using `prj_<build>.conf` scheme is aligned with upstream and thus we inherit this in rest of build system.

-----
Code snippet 1, that are node longer needed with this PR:
```
if (NOT CMAKE_BUILD_TYPE)
  set(CMAKE_BUILD_TYPE ZDebug)
endif()
-
if (NOT EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/configuration/${BOARD}")
  message(FATAL_ERROR
          "Board ${BOARD} is not supported.\n"
          "Please make sure board specific configuration files are added to "
          "${CMAKE_CURRENT_SOURCE_DIR}/configuration/${BOARD}")
endif()

if (NOT EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/configuration/${BOARD}/app_${CMAKE_BUILD_TYPE}.conf")
  message(FATAL_ERROR
          "Configuration file for build type ${CMAKE_BUILD_TYPE} is missing.\n"
          "Please add file ${CMAKE_CURRENT_SOURCE_DIR}/configuration/${BOARD}/app_${CMAKE_BUILD_TYPE}.conf")
endif()

if (EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/configuration/${BOARD}/dts.overlay")
  set(DTC_OVERLAY_FILE "${CMAKE_CURRENT_SOURCE_DIR}/configuration/${BOARD}/dts.overlay")
endif()

set(mcuboot_CONF_FILE
  ${CMAKE_CURRENT_SOURCE_DIR}/configuration/${BOARD}/mcuboot_${CMAKE_BUILD_TYPE}.conf
  )

if (EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/configuration/${BOARD}/spm_${CMAKE_BUILD_TYPE}.conf)
  set(spm_CONF_FILE
    ${CMAKE_CURRENT_SOURCE_DIR}/configuration/${BOARD}/spm_${CMAKE_BUILD_TYPE}.conf
    )
endif()

# Define configuration files.
set(CONF_FILE "configuration/${BOARD}/app_${CMAKE_BUILD_TYPE}.conf")
```

-------
Code snippet 2, that are node longer needed with this PR:
```
   if (EXISTS "configuration/${BOARD}/pm_static.yml"
        AND ${BOARD} STREQUAL "nrf52840dongle_nrf52840")
      set(PM_STATIC_YML_FILE configuration/${BOARD}/pm_static.yml
      )
    endif()
```

### CODE REVIEW

**Comment by tejlmand** [comment]
@hakonfam  @oyvindronningstad  those two commit are mainly of youir interrest: https://github.com/nrfconnect/sdk-nrf/pull/5117/commits/5f39a1aa29af60b563138dbe337b3a97014286f1, https://github.com/nrfconnect/sdk-nrf/pull/5117/commits/f6f66ec9e39a1a2bbaa9d8f6d754b5d214814d97.
Personally I find `pm_static_<board>_<revision>.yml` a bit long, so feel free to comment if the `pm_static_` part should have another name, maybe: `<board>_<revision>.pm.yml` ?
`<board>_<revision>` part is aligned on how we generally board specific conf files like those.

**Comment by rakons** [comment]
I have a problem with renames like this one: https://github.com/nrfconnect/sdk-nrf/pull/5117/files#diff-97994400eb6c7b41b21acb9bdaf75626f5e361490b76348a137c95255dec5b1f.

Why _ZRelease is renamed to _mbedtls?

**Review by rakons — APPROVED** [review]
Some doubts about some of the file renames - like in general commit.

File: cmake/boilerplate.cmake
Comment (carlescufi) [inline cmake/boilerplate.cmake:46]: Let us use `config` instead of `configuration`.

Comment (carlescufi) [reply cmake/boilerplate.cmake:46]: Alternative:
```
app/boards/nrf52840dk_nrf52840/{prj_release.conf,prj_debug.conf}
```

Comment (pdunaj) [reply cmake/boilerplate.cmake:46]: I am for "boards".
Earlier discussed it with @tejlmand. I think that it will fit nicely as all files in the directory are board specific application files.

**Comment by pdunaj** [comment]
I like the change. I expect the file names can be updated or we can do switch in a different PR. (we will definitly do a follow up as other apps will switch)
But I cannot build the image. Maybe it misses some local edit?

```
=== child image mcuboot -  begin ===
CMake Error at /home/pdunaj/work/ncs/nrf/cmake/extensions.cmake:130 (list):
  list sub-command REMOVE_ITEM requires two or more arguments.
```

File: cmake/boilerplate.cmake
Comment (mbolivar-nordic) [reply cmake/boilerplate.cmake:46]: > I am for "boards".

Seems reasonable. Would be nice to get this upstream though.

**Comment by tejlmand** [comment]
> I have a problem with renames like this one: https://github.com/nrfconnect/sdk-nrf/pull/5117/files#diff-97994400eb6c7b41b21acb9bdaf75626f5e361490b76348a137c95255dec5b1f.
> 
> Why _ZRelease is renamed to _mbedtls?

Agree, this looks wrong.
```
...ion/nrf52840dk_nrf52840/app_ZRelease.conf → ...tion/nrf52840dk_nrf52840/prj_mbedtls.conf 
```
an original change can be seen here:
https://github.com/pdunaj/fw-nrfconnect-nrf/pull/1/commits/316951a0d648e5f7931ba3d89d639b5b290d828f
```
...rf52840dk_nrf52840/app_ZDebugMbedTLS.conf → ...tion/nrf52840dk_nrf52840/prj_mbedtls.conf 
```
where the `app_ZDebugMbedTLS.conf` becomes `prj_mbedtls.conf`.

Will investigate if something went wrong on the rebase, or if it is a git presentation issue.

File: cmake/boilerplate.cmake
Comment (tejlmand) [reply cmake/boilerplate.cmake:46]: completely agree, so will do the change and if no side effects are noticed we can send an upstream patch.

**Comment by tejlmand** [comment]
> === child image mcuboot -  begin ===
CMake Error at /home/pdunaj/work/ncs/nrf/cmake/extensions.cmake:130 (list):
  list sub-command REMOVE_ITEM requires two or more arguments.

CMake version. CMake 3.20 is more forgiven on this and allows `list(REMOVE_ITEM ARGN )` (which happens when PREPROCESS_ARGS_KEYWORDS_MISSING_VALUES is empty), but seems CMake 3.13 complains.

Will fix.

**Comment by tejlmand** [comment]
> > === child image mcuboot -  begin ===
> > CMake Error at /home/pdunaj/work/ncs/nrf/cmake/extensions.cmake:130 (list):
> > list sub-command REMOVE_ITEM requires two or more arguments.
> 
> CMake version. CMake 3.20 is more forgiven on this and allows `list(REMOVE_ITEM ARGN )` (which happens when PREPROCESS_ARGS_KEYWORDS_MISSING_VALUES is empty), but seems CMake 3.13 complains.
> 
> Will fix.

and as one more info, `<prefix>_KEYWORDS_MISSING_VALUES` is new in 3.15, but as there are plans to bump CMake minimum required for improved toolchain support, then this is acceptable for now.

Just be aware that CMake >=3.15 is required for this PR.

**Comment by tejlmand** [comment]
> Signature file is found at right place but for some reason build system warns that default key is used...
Application fails to build complaining that /home/pdunaj/work/ncs/bootloader/mcuboot/mcuboot_private.pem cannot be found which would suggest that although child image gets the right key, app image still uses the default key.

I have found the reason for this.
This code:
https://github.com/nrfconnect/sdk-nrf/blob/c34ceebceb4fccd6e2fda1fc27c54640ca498f69/modules/mcuboot/CMakeLists.txt#L133-L139

will use the info from mcuboot_CONF_FILE, which were manually being defined here:
https://github.com/nrfconnect/sdk-nrf/blob/master/applications/pelion_client/CMakeLists.txt#L31-L33

But that is no longer the case when we apply the automatic child image configuration scheme.
That seems like a flaw / missing part / process order dependency in other parts of the build infrastructure that must be corrected.

Currently unsure why it's not being properly applied.

Looking into a fix.

**Comment by tejlmand** [comment]
> Currently unsure why it's not being properly applied.
> 
> Looking into a fix.

@pdunaj  fixed.
Note, mcuboot expects the pem file in same folder as the mcuboot_CONF_FILE, which is the reason why the pem files has been relocated.

I think it would make sense to reconsider that requirement, but would like to take such discussion outside this PR itself.

**Comment by pdunaj** [comment]
> > Currently unsure why it's not being properly applied.
> > Looking into a fix.
> 
> @pdunaj fixed.
> Note, mcuboot expects the pem file in same folder as the mcuboot_CONF_FILE, which is the reason why the pem files has been relocated.
> 
> I think it would make sense to reconsider that requirement, but would like to take such discussion outside this PR itself.

Yes, I moved the file during testing.
Thanks for fixing. Will give it a go and see if something is wrong.

**Comment by tejlmand** [comment]
> > Why _ZRelease is renamed to _mbedtls?
> 
> Agree, this looks wrong.

@rakons, should be fixed now.

**Review by pdunaj — CHANGES_REQUESTED** [review]
Still fails to find `mcuboot_private.pem`.

**Review by pdunaj — APPROVED** [review]
it seems I have not rebase :) the change works
I have added the patch cleaning up renames in pelion_client

File: cmake/boilerplate.cmake
Comment (mbolivar-nordic) [reply cmake/boilerplate.cmake:46]: > completely agree, so will do the change and if no side effects are noticed we can send an upstream patch.

My feeling is we should try to upstream this before merging it downstream if possible, but I won't block over it.

Comment (pdunaj) [reply cmake/boilerplate.cmake:46]: @mbolivar-nordic thanks. I would like to have it cleaned up for NCS1.7 but for that to happen we need to finish the whole rework before mid August. I think it's not the best idea to do this change in rush around the release.
If the change cannot enter upstream quickly (maybe it can, I don't know), I would recommend to have it in downstream asap. Once it's in we would need to switch machine learning and nrf_desktop. In all cases we need to update the documentation. All of this takes time.

**Review by sebastiandraus — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
> it seems I have not rebase :) the change works

Are we expecting green CI? It's currently red.

**Comment by pdunaj** [comment]
Rebased

**Comment by pdunaj** [comment]
> Are we expecting green CI? It's currently red.

The CI was green when the last commit was pushed about a week ago. I expect it to be green again but let's see.

**Comment by pdunaj** [comment]
Fails on word check.

> T5 Title contains the word 'wip'

**Comment by mbolivar-nordic** [comment]
> Fails on word check.
> 
> > T5 Title contains the word 'wip'

Indeed, and the commit with wip in it says:

` wip: cmake: CMake >=3.15 is required for this PR to work `

It seems we need to update the cmake requirement for NCS before we can merge this.

File: cmake/extensions.cmake
Comment (hakonfam) [inline cmake/extensions.cmake:172]: IMO the name scheme proposed here is good.

**Comment by MarcusCHZ** [comment]
This makes configuration in line with other apps.

**Comment by tejlmand** [comment]
> `wip: cmake: CMake >=3.15 is required for this PR to work`
> 
> It seems we need to update the cmake requirement for NCS before we can merge this.

upstream PR bumping CMake version to 3.20 is now in place:
https://github.com/zephyrproject-rtos/zephyr/pull/37575

**Comment by pdunaj** [comment]
Hi @tejlmand , do you think it is doable to have this merged this week? If we want to rework all our apps we need some stabilization time for everything that can go wrong to show itself before the release.

**Comment by carlescufi** [comment]
@mbolivar-nordic @tejlmand do we need to wait until the current upmerge is complete to merge this?

**Comment by mbolivar-nordic** [comment]
@carlescufi aren't we waiting for https://github.com/nrfconnect/sdk-nrf/pull/5262 and https://github.com/zephyrproject-rtos/zephyr/pull/37575?

File: cmake/boilerplate.cmake
Comment (tejlmand) [reply cmake/boilerplate.cmake]: I now have done the changes here:
https://github.com/nrfconnect/sdk-zephyr/pull/580/commits/ce90123da29ae723fdc55063dbc416476e12ec6c

@carlescufi @mbolivar-nordic could one of you take a quick look and see if this is in line with your expectations.
If yes, I will make an upstream PR and change title to `[nrf fromlist]`

**Comment by tejlmand** [comment]
rebased and updated.

Please take a look.

File: cmake/boilerplate.cmake
Comment (mbolivar-nordic) [reply cmake/boilerplate.cmake]: > @carlescufi @mbolivar-nordic could one of you take a quick look

I only read the commit message, but it like a good enough proposal to open a PR upstream.

**Comment by pdunaj** [comment]
I propose to merge this after the release. Risk of something going wrong is quite high now, there is little time to update docs plus we won't be able to sync all our apps so there will be mess anyway.
We can merge it right after ncs1.7 is out.

**Comment by mbolivar-nordic** [comment]
> We can merge it right after ncs1.7 is out.

The argument to merge it downstream before getting it upstream previously seemed to be "we want this for 1.7". But now it's not going into 1.7 because it depends on the CMake update which is still not in. So now I'm wondering why there's a rush to get it in before the upstream PR has a chance to get reviewed and merged.

In general divergences at this level in NCS are harmful if Zephyr adopts a different mechanism. Already we field many confused users on the Zephyr slack who do not understand the difference between upstream and NCS. Partition manager is the usual culprit and I don't think we should make it worse without good reason.

**Comment by pdunaj** [comment]
> > We can merge it right after ncs1.7 is out.
> 
> The argument to merge it downstream before getting it upstream previously seemed to be "we want this for 1.7". But now it's not going into 1.7 because it depends on the CMake update which is still not in. So now I'm wondering why there's a rush to get it in before the upstream PR has a chance to get reviewed and merged.
> 
> In general, divergences at this level in NCS are harmful if Zephyr adopts a different mechanism. Already we field many confused users on the Zephyr slack who do not understand the difference between upstream and NCS. Partition manager / multi-image is the usual culprit, and I don't think we should make it worse without good reason.

I would stay on the path. We should get it ASAP after the release so we can start working on other application configuration. We cannot wait for the last minute of the release. This change is simply too big and important.

**Comment by tejlmand** [comment]
>     In general, divergences at this level in NCS are harmful if Zephyr adopts a different mechanism. Already we field many confused users on the Zephyr slack who do not understand the difference between upstream and NCS. Partition manager / multi-image is the usual culprit, and I don't think we should make it worse without good reason.
>
> I would stay on the path. We should get it ASAP after the release so we can start working on other application configuration. We cannot wait for the last minute of the release. This change is simply too big and important.

what @mbolivar-nordic refers to is not the solution as such and that we cannot merge it early, but that if `upstream` decides the `<app>/boards/<board>/` is not a good default location, but they want `<app>/something-else/<board>/` is better and we have moved all our board configs to the initial proposal then we start to deviate from what upstream does.

This is a very user faced change, so that's why some initial feedback upstream had been nice, so we could see if this proposal was being accepted with suggested default location.
https://github.com/zephyrproject-rtos/zephyr/pull/38229

**Comment by pdunaj** [comment]
Working on rebase. Force pushed zephyr side already - seen no conflicts there.
The problem I see is with partition manager ignoring perlion_storage. It simply throws it away :/

**Comment by pdunaj** [comment]
Next problem. Cannot change build type...
```
CMake Error at /home/pdunaj/work/ncs/zephyr/cmake/kconfig.cmake:209 (message):
  File not found:
  /home/pdunaj/work/ncs/nrf/applications/pelion_client/prj_release.conf
```

**Comment by pdunaj** [comment]
Rebased but this still does not handle nrf91 well - partition's are wrong again.

**Review by pdunaj — CHANGES_REQUESTED** [review]
Rebase in progress. Still problems with nrf91

**Comment by pdunaj** [comment]
Will push the fix tomorrow along with rebase on master.

Hi @tejlmand , do you think we can get it in quickly?

**Comment by pdunaj** [comment]
I have pushed today and I think I am done with rebase. Will check if the dotconfig does not change for affected configurations and approve once it's done.

**Comment by pdunaj** [comment]
Doc CI failed as kconfig options changed. Let's see if small fix makes it green.

**Comment by pdunaj** [comment]
```
[2021-09-29T07:38:02.672Z] FileNotFoundError: [Errno 2] No such file or directory: '/jenkins_workspace/workspace/b_test-fw-nrfconnect-apps_master/test_objects/ncs/nrf/applications/pelion_client/configuration/common/update_default_resources.c'
```
When we will be right before merge we should switch the tests (@piwe-nordic , @jumi-nordic ).

Other failures seems unconnected.

**Review by FrancescoSer — APPROVED** [review]

**Review by simensrostad — APPROVED** [review]

File: cmake/boilerplate.cmake
Comment (tejlmand) [reply cmake/boilerplate.cmake]: > In a private conversation @tejlmand expressed that we should not use boards for this purpose.

Correct, cause I feel that this messes up the boards folder in the app, and the description originates from the days before we actually used the boards folder for dt overlays and Kconfig fragments.
Now we have Zephyr modules in place which support a module.yml and thus easy integration of custom board to the project with automatic broader availability then I think this doc should be updated.

But following the referred docs will not impact whatever choices we make in this PR.

**Comment by wbober** [comment]
> > [#5117 (comment)](https://github.com/nrfconnect/sdk-nrf/pull/5117#discussion_r730619075)
> 
> Not sure what kind of changes you are requesting ?

@tejlmand No changes just answers :) I've abused the review to makes sure that we are aligned on this.

> Now we have Zephyr modules in place which support a module.yml and thus easy integration of custom board to the project with automatic broader availability then I think this doc should be updated.

But is this what we are going to recommend from now on? From my talks with Ingar and looking at the [example application in zephyr](https://github.com/zephyrproject-rtos/example-application) it seems to me that putting board files next to the app code is a valid use case.
 
> Not sure what the custom board implementation has to do with this ? Users with a custom board root must still specify this using `-DBOARD_ROOT=<path>` or set it directly in CMake with `set(BOARD_ROOT ...)`

Correct. And I'd like to modify this (clearly not in this PR). The build system already does quite a bit to figure out correct configuration files. I think it won't be much of an overkill if it automatically adds `BOARD_ROOT` if it exists. I'd just like to make sure that this PR here doesn't conflict with what I'd like to achieve.

**Pull request merged**
